processor/src/main/java/org/mapstruct/ap/internal/processor/MethodRetrievalProcessor.java

```diff
@@ -406,7 +406,13 @@ private boolean hasFactoryAnnotation(ExecutableElement method) {
         return ObjectFactoryGem.instanceOn( method ) != null;
     }
 
-    private boolean isValidPresenceCheckMethod(ExecutableElement method, Type returnType) {
+    private boolean isValidPresenceCheckMethod(ExecutableElement method, List<Parameter> parameters, Type returnType) {
+        for ( Parameter param : parameters ) {
+            if ( param.isTargetPropertyName() && !param.getType().isString() ) {
+                messager.printMessage( method, Message.RETRIEVAL_TARGET_PROPERTY_NAME_WRONG_TYPE );
```

Comment:
We can actually pass the annotation mirror from the `@TargetPropertyName` here as well. We also should use the parameter name element for the error element.

e.g.

```java
message.printMessage(
    param.getElement(),
    TargetPropertyNameGem.instanceOn( param.getElement() ).mirror() ),
    Message. RETRIEVAL_TARGET_PROPERTY_NAME_WRONG_TYPE
);
```